ozark/src/main/java/org/glassfish/ozark/locale/LocaleResolverChain.java

```diff
@@ -0,0 +1,126 @@
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package org.glassfish.ozark.locale;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.Priority;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+import javax.mvc.locale.LocaleResolver;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Configuration;
+import javax.ws.rs.core.Context;
+import java.lang.annotation.Annotation;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Implements the locale resolving algorithm described in {@link LocaleResolver}.
+ *
+ * @author Christian Kaltepoth
+ */
+@ApplicationScoped
+public class LocaleResolverChain {
+
+    @Inject
+    @Any
+    private Instance<LocaleResolver> resolvers;
+
+    @Inject
+    private HttpServletRequest request;
+
+    @Context
+    private Configuration configuration;
+
+    @PostConstruct
+    public void verify() {
+        Objects.requireNonNull(configuration, "The Configuration instance was not injected! " +
+                "Please make sure you are using a recent version of Jersey.");
+    }
+
+    public Locale resolve() {
+
+        // prepare context instance
+        LocaleResolverContextImpl context = new LocaleResolverContextImpl();
+        context.setConfiguration(configuration);
+        context.setRequest(request);
+
+        // candidates as sorted list
+        List<LocaleResolver> candidates = StreamSupport.stream(resolvers.spliterator(), false)
+                .sorted((resolver1, resolver2) -> {
+                    final Priority prio1 = getAnnotation(resolver1.getClass(), Priority.class);
+                    final Priority prio2 = getAnnotation(resolver2.getClass(), Priority.class);
+                    final int value1 = prio1 != null ? prio1.value() : 1000;
+                    final int value2 = prio2 != null ? prio2.value() : 1000;
+                    return value2 - value1;
+                })
+                .collect(Collectors.toList());
+
+        // do the resolving
+        for (LocaleResolver candidate : candidates) {
+            Locale locale = candidate.resolveLocale(context);
+            if (locale != null) {
+                return locale;
+            }
+        }
+
+        throw new IllegalStateException("Could not resolve locale with any of the " + candidates.size()
+                + " resolver implementations");
+
```

Comment:
it may be a bit more human readable extracting a few methods and exploring a bit more the streams api:
```
StreamSupport.stream(resolvers.spliterator(), false)
    .sorted(this::byPriority)
    .map(candidate -> candidate.resolveLocale(context))
    .filter(candidate -> candidate != null)
    .findFirst()
    .orElseThrow(() -> new IllegalStateException("Could not resolve ..."));
```
what do you think?

note: see my comments about optional on mvc-spec https://github.com/mvc-spec/mvc-spec/pull/2/files#r58268281